xcm/xcm-builder/src/fee_handling.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::marker::PhantomData;
+use frame_support::traits::{Contains, Get};
+use xcm::prelude::*;
+use xcm_executor::traits::{FeeManager, FeeReason, TransactAsset};
+
+/// A `FeeManager` implementation that simply deposits the fees handled into a specific on-chain
+/// `ReceiverAccount`.
+///
+/// It reuses the `AssetTransactor` configured on the XCM executor to deposit fee assets, and also
+/// permits specifying `WaivedLocations` for locations that are privileged to not pay for fees.
+pub struct XcmFeesToAccount<XcmConfig, WaivedLocations, AccountId, ReceiverAccount>(
+	PhantomData<(XcmConfig, WaivedLocations, AccountId, ReceiverAccount)>,
+);
+impl<
+		XcmConfig: xcm_executor::Config,
+		WaivedLocations: Contains<MultiLocation>,
+		AccountId: Clone + Into<[u8; 32]>,
+		ReceiverAccount: Get<Option<AccountId>>,
+	> FeeManager for XcmFeesToAccount<XcmConfig, WaivedLocations, AccountId, ReceiverAccount>
+{
+	fn is_waived(origin: Option<&MultiLocation>, _: FeeReason) -> bool {
+		let Some(loc) = origin else { return false };
+		WaivedLocations::contains(loc)
+	}
+
+	fn handle_fee(fees: MultiAssets, context: Option<&XcmContext>) {
+		if let Some(receiver) = ReceiverAccount::get() {
+			let dest = AccountId32 { network: None, id: receiver.into() }.into();
+			for asset in fees.into_inner() {
+				let ok = XcmConfig::AssetTransactor::deposit_asset(&asset, &dest, context).is_ok();
```

Comment:
I think, this could fail, if ReceiverAccount does not have at least existentional deposit

Reply:
If you are really sure, then I think we should document this on the struct itself.

Reply:
maybe, depends on underlying transactor impl,
but let me check/try/verify tomorrow

Reply:
If the method return type is result, you must assume it can fail. This is library code that should work with any implementation of the transactor.
So the debug_assert is not correct.

Reply:
I don't see why it's correct to conclude the `debug_assert` is not correct. For the purposes of this particular functionality, it IS crucial to ensure that the deposit works as intended, otherwise the funds will get burnt. Hence, any implementation of the asset transactor needs to guarantee this, otherwise we should have a `debug_assert` to warn implementers of a bad implementation.

Reply:
This is leaky abstraction. I have an implementation of asset transactor with orml-tokens and I can not guarantee this. What should I do?

Reply:
Then the funds will get burnt. I can only think of putting it into an asset trap otherwise, claimable only by the specified account. If this isn't fine, then tell us how else we can handle the case where the asset transactor fails depositing the asset.

Reply:
just let it burn and document it

Reply:
my point is, don’t make false assumptions it cannot fail. don’t make infeasible requirement it cannot fail. just accept it can fail and handle it properly

Reply:
or maybe also log it like:
```
                        if let Err(e) = XcmConfig::AssetTransactor::deposit_asset(&asset, &dest, context) {
				log::warn!("xcm:fees", "Burning fees: {:?}, because of `AssetTransactor::deposit_asset` error: {:?}", asset, e);
			}

```

Reply:
Maybe a warn rather than an assert - that way we get to see that in production code too? (and given that the asset transactor implementation is pluggable)